File: kattis/wheresmyinternet/wheresmyinternet.cpp
#include <iostream>

using namespace std;

struct UnionFind {
  int parent[200001];
  
  void initialise(int n) 
  {
    for (int i = 1; i < n + 1; i++)
      parent[i] = i;
  }

  void _union(int a, int b)
  {
    a = find(a);
    b = find(b);
    if (a == b) return;
    
    if (b < a) {
      int temp = a;
      a = b;
      b = temp;
    }

    parent[b] = a;
  }

  int find(int a) 
  {
    while (a != parent[a]) {
      parent[a] = parent[parent[a]];
      a = parent[a];
    }
    return a;
  }
};


int main()
{
  int n, m, a, b, i;
  UnionFind network;
  cin >> n >> m;
  network.initialise(n);
  while (m--) {
    cin >> a >> b;
    network._union(a, b);
  }

  bool possible = true;
  for (i = 2; i < n + 1; i++) {
    if (network.find(i) != 1) {
      cout << i << "\n";
      possible = false;
    }
  }

  if (possible) cout << "Connected";
  return 0;
}

File: kattis/wheresmyinternet/wheresmyinternet.py
#!/usr/bin/env python3


class UnionFind(object):
    def __init__(self):
        self.unions = {}

    def add_node(self, a):
        self.unions[a] = a
    
    def find(self, a):
        while a != self.unions[a]:
            self.unions[a] = self.unions[self.unions[a]]
            a = self.unions[a]
        return a

    def union(self, a, b):
        a_parent = self.find(a)
        b_parent = self.find(b)
        if a_parent == "1":
            self.unions[b_parent] = a_parent
        else:
            self.unions[a_parent] = b_parent


network = UnionFind()
m, n = map(int, input().split())

for i in range(n):
    a, b = input().split()
    if a not in network.unions:
        network.add_node(a)
    if b not in network.unions:
        network.add_node(b)
    network.union(a, b)

connected = 1
for i in range(2, m + 1):
    i = str(i)
    if i not in network.unions or network.find(i) != "1":
        print(i)
        connected = 0

if connected:
    print("Connected")
